Fix missing includes and guards when certain features are disabled

spenap opened the pull request:
#### 951a118e17cd92adc3aa31129af7cd348fac91e9
<pre>
Fix missing includes and guards when certain features are disabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=284523">https://bugs.webkit.org/show_bug.cgi?id=284523</a>

Reviewed by Philippe Normand.

When disabling long-standing features such as VIDEO, WEB_AUDIO, WEB_CODECS
and others, WebKit failed to build as some of the symbols were not
properly guarded, or includes were missing (probably because they were
transitively included otherwise).

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::privateBrowsingStateDidChange): Missing UNUSED_PARAM
when VIDEO is not enabled.
* Source/WebCore/loader/LinkLoader.cpp:
(WebCore::LinkLoader::resourceTypeFromAsAttribute): TextTrackResource
can&apos;t be returned with VIDEO not enabled.
(WebCore::createLinkPreloadResourceClient): Skip TextTrackResource with
VIDEO not enabled.
(WebCore::LinkLoader::isSupportedType): ditto.
* Source/WebCore/loader/SubresourceLoader.cpp:
(WebCore::logResourceLoaded): ditto.
* Source/WebCore/loader/cache/CachedResource.h: ditto.
* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::createResource): ditto.
* Source/WebCore/loader/cache/CachedResourceLoader.h: Do not expose
  CachedTextTrack if VIDEO is not enabled.
* Source/WebCore/loader/cache/CachedTextTrack.cpp: Do not expose if
  VIDEO is not enabled.
* Source/WebCore/loader/cache/CachedTextTrack.h: Do not expose if VIDEO
  is not enabled.
* Source/WebCore/platform/audio/PlatformMediaSessionManager.cpp:
(WebCore::PlatformMediaSessionManager::endInterruption): Missing
UNUSED_PARAM when VIDEO is not enabled.
(WebCore::PlatformMediaSessionManager::addSession): Protect
Session-specific members when VIDEO and WEB_AUDIO are not enabled.
(WebCore::PlatformMediaSessionManager::removeSession): ditto.
(WebCore::PlatformMediaSessionManager::sessionWillEndPlayback): ditto.
(WebCore::PlatformMediaSessionManager::setCurrentSession): ditto.
(WebCore::PlatformMediaSessionManager::applicationWillBecomeInactive):
ditto.
(WebCore::PlatformMediaSessionManager::applicationDidBecomeActive):
ditto.
(WebCore::PlatformMediaSessionManager::applicationDidEnterBackground):
ditto.
(WebCore::PlatformMediaSessionManager::applicationWillEnterForeground):
ditto.
(WebCore::PlatformMediaSessionManager::processWillSuspend): ditto.
(WebCore::PlatformMediaSessionManager::processDidResume): ditto.
(WebCore::PlatformMediaSessionManager::processDidReceiveRemoteControlCommand): ditto.
(WebCore::PlatformMediaSessionManager::processSystemWillSleep): ditto.
(WebCore::PlatformMediaSessionManager::processSystemDidWake): ditto.
(WebCore::PlatformMediaSessionManager::stopAllMediaPlaybackForProcess):
ditto.
(WebCore::PlatformMediaSessionManager::suspendAllMediaPlaybackForGroup):
ditto.
(WebCore::PlatformMediaSessionManager::resumeAllMediaPlaybackForGroup):
ditto.
(WebCore::PlatformMediaSessionManager::bestEligibleSessionForRemoteControls): ditto.
(WebCore::PlatformMediaSessionManager::hasActiveNowPlayingSessionInGroup): ditto.
(WebCore::PlatformMediaSessionManager::dumpSessionStates): ditto.
* Source/WebCore/platform/audio/PlatformRawAudioData.cpp: Do not expose
  if WEB_CODECS are not enabled.
* Source/WebCore/platform/audio/PlatformRawAudioData.h: ditto.
* Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp: Missing include.
* Source/WebKit/UIProcess/API/glib/WebKitProtocolHandler.cpp:
(WebKit::WebKitProtocolHandler::handleGPU): Protected renderMediaBuffer
usage with USE(LIBDRM)
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp: Guard
  MHTML-related logic better, and fail saving if unavailable.
* Source/WebKit/UIProcess/Automation/libwpe/WebAutomationSessionWPE.cpp:
  Do not expose unused methods when WebDriver&apos;s Mouse, Wheel or Touch
interactions are disabled.
* Source/WebKit/UIProcess/wpe/WebContextMenuProxyWPE.h: Do not expose
  when CONTEXT_MENUS is not enabled.
* Source/WebKit/WebProcess/GPU/ShapeDetection/RemoteBarcodeDetectorProxy.h:
  Moved the guard after the type is defined
* Source/WebKit/WebProcess/GPU/ShapeDetection/RemoteFaceDetectorProxy.h:
  ditto.
* Source/WebKit/WebProcess/GPU/ShapeDetection/RemoteTextDetectorProxy.h:
  ditto.
* Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp:
(WebKit::maximumBufferingTime): Do not expose TextTrackResource if VIDEO
is not enabled.
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp: Missing FrameLoader
  include.
* Source/WebKit/WebProcess/WebPage/WebPage.h: Missing RegistrableDomain
  include.

Canonical link: <a href="https://commits.webkit.org/288522@main">https://commits.webkit.org/288522@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/14a18c2cafb34fb3655a909cebb640d4bcae342d

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/83693 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/3310 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/37994 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/88765 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/34701 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/85778 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/3400 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/11268 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65103 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/22935 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/86739 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/2510 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/76045 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45391 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/2438 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/30255 "Found 1 new test failure: fast/editing/recursive-reapply-edit-command-crash.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/33750 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/73472 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/31001 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/90143 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/10958 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/7911 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/73539 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/11182 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/71870 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/72763 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17990 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17025 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/15724 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/2282 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/10910 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/16382 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/10758 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/14233 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/12530 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

TingPing commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -4757,9 +4757,9 @@ struct ViewSaveAsyncData {
     RefPtr<API::Data> webData;
     GRefPtr<GFile> file;
 };
+#if ENABLE(MHTML)
```
I do think `ViewSaveAsyncData` belongs behind this also.

I see `webkit_web_view_save_finish()` uses this type not behind a guard, but that doesn't make sense since `webkit_web_view_save()` is guarded, so it will never work.

`webkit_web_view_save()` probably should have been setting an error instead of doing nothing, but in practice this situation probably isn't often hit.

spenap replied:
I agree with your comments, and have updated the code so that when `MHTML` is not enabled it does `g_task_return_boolean(task, FALSE);`. 

I was tempted to return an error, but seeing that errors need registering, I am not sure it's worth as `MHTML` will remain enabled outside experiments. Do you think that's acceptable?

TingPing replied:
I'd use a generic error, like `g_task_return_new_error_literal(..., G_IO_ERROR, G_IO_ERROR_NOT_SUPPORTED, "WebKit was built without MHTML support which this API requires");`

philn commented on Source/WebCore/platform/audio/PlatformMediaSessionManager.cpp:
```diff
@@ -362,7 +370,9 @@ void PlatformMediaSessionManager::sessionStateChanged(PlatformMediaSession& sess
 
 void PlatformMediaSessionManager::setCurrentSession(PlatformMediaSession& session)
 {
+#if ENABLE(VIDEO) || ENABLE(AUDIO)
```
There's no `ENABLE(AUDIO)`, should be `WEB_AUDIO` here and below.

spenap committed:
Fix missing includes and guards when certain features are disabled
https://bugs.webkit.org/show_bug.cgi?id=284523

Reviewed by Philippe Normand.

When disabling long-standing features such as VIDEO, WEB_AUDIO, WEB_CODECS
and others, WebKit failed to build as some of the symbols were not
properly guarded, or includes were missing (probably because they were
transitively included otherwise).

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::privateBrowsingStateDidChange): Missing UNUSED_PARAM
when VIDEO is not enabled.
* Source/WebCore/loader/LinkLoader.cpp:
(WebCore::LinkLoader::resourceTypeFromAsAttribute): TextTrackResource
can't be returned with VIDEO not enabled.
(WebCore::createLinkPreloadResourceClient): Skip TextTrackResource with
VIDEO not enabled.
(WebCore::LinkLoader::isSupportedType): ditto.
* Source/WebCore/loader/SubresourceLoader.cpp:
(WebCore::logResourceLoaded): ditto.
* Source/WebCore/loader/cache/CachedResource.h: ditto.
* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::createResource): ditto.
* Source/WebCore/loader/cache/CachedResourceLoader.h: Do not expose
  CachedTextTrack if VIDEO is not enabled.
* Source/WebCore/loader/cache/CachedTextTrack.cpp: Do not expose if
  VIDEO is not enabled.
* Source/WebCore/loader/cache/CachedTextTrack.h: Do not expose if VIDEO
  is not enabled.
* Source/WebCore/platform/audio/PlatformMediaSessionManager.cpp:
(WebCore::PlatformMediaSessionManager::endInterruption): Missing
UNUSED_PARAM when VIDEO is not enabled.
(WebCore::PlatformMediaSessionManager::addSession): Protect
Session-specific members when VIDEO and WEB_AUDIO are not enabled.
(WebCore::PlatformMediaSessionManager::removeSession): ditto.
(WebCore::PlatformMediaSessionManager::sessionWillEndPlayback): ditto.
(WebCore::PlatformMediaSessionManager::setCurrentSession): ditto.
(WebCore::PlatformMediaSessionManager::applicationWillBecomeInactive):
ditto.
(WebCore::PlatformMediaSessionManager::applicationDidBecomeActive):
ditto.
(WebCore::PlatformMediaSessionManager::applicationDidEnterBackground):
ditto.
(WebCore::PlatformMediaSessionManager::applicationWillEnterForeground):
ditto.
(WebCore::PlatformMediaSessionManager::processWillSuspend): ditto.
(WebCore::PlatformMediaSessionManager::processDidResume): ditto.
(WebCore::PlatformMediaSessionManager::processDidReceiveRemoteControlCommand): ditto.
(WebCore::PlatformMediaSessionManager::processSystemWillSleep): ditto.
(WebCore::PlatformMediaSessionManager::processSystemDidWake): ditto.
(WebCore::PlatformMediaSessionManager::stopAllMediaPlaybackForProcess):
ditto.
(WebCore::PlatformMediaSessionManager::suspendAllMediaPlaybackForGroup):
ditto.
(WebCore::PlatformMediaSessionManager::resumeAllMediaPlaybackForGroup):
ditto.
(WebCore::PlatformMediaSessionManager::bestEligibleSessionForRemoteControls): ditto.
(WebCore::PlatformMediaSessionManager::hasActiveNowPlayingSessionInGroup): ditto.
(WebCore::PlatformMediaSessionManager::dumpSessionStates): ditto.
* Source/WebCore/platform/audio/PlatformRawAudioData.cpp: Do not expose
  if WEB_CODECS are not enabled.
* Source/WebCore/platform/audio/PlatformRawAudioData.h: ditto.
* Source/WebKit/Platform/IPC/unix/ConnectionUnix.cpp: Missing include.
* Source/WebKit/UIProcess/API/glib/WebKitProtocolHandler.cpp:
(WebKit::WebKitProtocolHandler::handleGPU): Protected renderMediaBuffer
usage with USE(LIBDRM)
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp: Guard
  MHTML-related logic better, and fail saving if unavailable.
* Source/WebKit/UIProcess/Automation/libwpe/WebAutomationSessionWPE.cpp:
  Do not expose unused methods when WebDriver's Mouse, Wheel or Touch
interactions are disabled.
* Source/WebKit/UIProcess/wpe/WebContextMenuProxyWPE.h: Do not expose
  when CONTEXT_MENUS is not enabled.
* Source/WebKit/WebProcess/GPU/ShapeDetection/RemoteBarcodeDetectorProxy.h:
  Moved the guard after the type is defined
* Source/WebKit/WebProcess/GPU/ShapeDetection/RemoteFaceDetectorProxy.h:
  ditto.
* Source/WebKit/WebProcess/GPU/ShapeDetection/RemoteTextDetectorProxy.h:
  ditto.
* Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp:
(WebKit::maximumBufferingTime): Do not expose TextTrackResource if VIDEO
is not enabled.
* Source/WebKit/WebProcess/WebPage/WebFrame.cpp: Missing FrameLoader
  include.
* Source/WebKit/WebProcess/WebPage/WebPage.h: Missing RegistrableDomain
  include.

Canonical link: https://commits.webkit.org/288522@main
